oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+allow for the ability to dynamically add CSS and JavaScript. See their blog
+post `How we switched our template rendering engine to React`_ for more
+details.
+
+Another similar use case is HTML5 ads, but those are all implemented as
+iframes. Using iframes is possible for some pluggable situations (such as LTI),
+but it would provide a poor user experience if a complex page was constructed
+from tens or even hundreds of iframes.
+
+Could we use XBlocks instead of introducing Django Component Views?
+-------------------------------------------------------------------
+
+Note that all of the requirements of pluggable user interfaces are supported by
+XBlocks. However, there are several reasons why XBlocks are not always
+appropriate:
+
+* XBlocks need to have a unique usage id so that instances can be referenced.
+
+* XBlocks need to have storage scoped for different purposes (per learner, per
+  course etc).
+
+* XBlocks in the LMS are tied to a course, which doesn't make them great for
+  general user interfaces (e.g. UIs that are global, organization-wide or
+  program-wide).
+
+* XBlocks are persisted while component views are ephemeral. This also means
+  that converting an existing view into an XBlock requires a data migration
+  so that each course is updated to contain a block instance.
+
+* XBlocks need to support serialization to OLX while component views don't.
+
+* XBlocks are designed to be platform-agnostic and so in a pure implementation
+  should only fetch data through provided runtime services. This has been a
+  stumbling block for adding new XBlocks as the data is usually only made
+  available through REST APIs.
+
+Will Django Component Views ever need to have their own storage?
+----------------------------------------------------------------
+
+The intention is that Django Component Views provide new user interfaces, and so
+should not have any storage of their own. They will need to be provided with
+context (e.g. the course that is being rendered) and they will need to be able
+to fetch data (e.g. the children of an xblock that they are rendering). If a new
+feature needs storage of i  ts own, then that should be implemented as a Django
+model. Django views do not provide storage mechanisms, and Django component
+views, as subclasses of Django views, should not either.
+
+Should Django Component Views support nesting?
+----------------------------------------------
+
+Django Component Views should not be constructed as a tree of views as that is
+introducing state that needs to be managed. However, component views can invoke
+other component views to render portions of their content. For example, a Unit
+Student View could ask each of its children to render itself as a fragment, and
+then could return a composite fragment containing all of them. This is
+essentially how XBlocks render their children today, and component views should
+work in the same way.
+
```

Comment:
I'd like to avoid arbitrary nesting, or at least discourage it until we've proven that we can have it working well in an async manner. One of the biggest performance issue with XBlocks has always been the blocking behavior of having to invoke arbitrary children and grandchildren, and not being able to return until they've all executed.